goto-cc fails to parse enum/struct/unions with C23 attribute sequences

For example, the following code fails to parse:

```c
enum [[nodiscard]] error_t {
  A,
};

int main() {
}
```

The error output is:

```
test.c:2:1: error: syntax error before '['
 enum [[nodiscard]] error_t {
PARSING ERROR
```

This also fails in the exact same way for structs or unions with an attribute sequence.

--- a/regression/ansi-c/c23_attributes1/main.c
+++ b/regression/ansi-c/c23_attributes1/main.c
@@ -1,0 +1,8 @@
+enum [[nodiscard]] error_t
+{
+  A,
+};
+
+int main()
+{
+}

--- a/regression/ansi-c/c23_attributes1/test.desc
+++ b/regression/ansi-c/c23_attributes1/test.desc
@@ -1,0 +1,8 @@
+CORE
+main.c
+-std=c2x
+^EXIT=0$
+^SIGNAL=0$
+--
+^warning: ignoring
+^CONVERSION ERROR$

--- a/src/ansi-c/scanner.l
+++ b/src/ansi-c/scanner.l
@@ -899,8 +899,18 @@ enable_or_disable ("enable"|"disable")
 "__if_not_exists"   { return MSC_cpp_keyword(TOK_MSC_IF_NOT_EXISTS); }
 "__underlying_type" { return conditional_keyword(PARSER.cpp98, TOK_UNDERLYING_TYPE); }
 
-"[["                { if(PARSER.c23)
+"[["                { // C23 attributes (also C++11 and later, but for C++ we
+                      // handle them directly in the parser); GCC >= 11, Clang
+                      // >= 17, and Visual Studio >= 2022 support these
+                      // irrespective of the language standard selected on the
+                      // command-line.
+                      if(PARSER.c23 ||
+                         PARSER.mode==configt::ansi_ct::flavourt::GCC ||
+                         PARSER.mode==configt::ansi_ct::flavourt::CLANG ||
+                         PARSER.mode==configt::ansi_ct::flavourt::VISUAL_STUDIO)
+                      {
                         BEGIN(STD_ANNOTATION);
+                      }
                       else
                       {
                         yyless(1); // puts one [ back into stream

--- a/src/goto-cc/gcc_mode.cpp
+++ b/src/goto-cc/gcc_mode.cpp
@@ -648,6 +648,20 @@ int gcc_modet::doit()
        std_string=="gnu1x" || std_string=="c1x")
       config.ansi_c.set_c11();
 
+    if(
+      std_string == "gnu17" || std_string == "c17" || std_string == "gnu18" ||
+      std_string == "c18")
+    {
+      config.ansi_c.set_c17();
+    }
+
+    if(
+      std_string == "gnu2x" || std_string == "c2x" || std_string == "gnu23" ||
+      std_string == "c23")
+    {
+      config.ansi_c.set_c23();
+    }
+
     if(std_string=="c++11" || std_string=="c++1x" ||
        std_string=="gnu++11" || std_string=="gnu++1x" ||
        std_string=="c++1y" ||

